examples/triplet/convert_3d_triplet_data.cpp

```diff
@@ -0,0 +1,163 @@
+#include <string>
+#include "caffe/proto/caffe.pb.h"
+#include "caffe/util/math_functions.hpp"
+#include "glog/logging.h"
+#include "google/protobuf/text_format.h"
+#include "leveldb/db.h"
+#include "stdint.h"
+
+uint32_t swap_endian(uint32_t val) {
+    val = ((val << 8) & 0xFF00FF00) | ((val >> 8) & 0xFF00FF);
+    return (val << 16) | (val >> 16);
+}
+
+void read_image(std::ifstream* image_file, std::ifstream* label_file,
+        uint32_t index, uint32_t rows, uint32_t cols,
+        char* pixels, char* label_temp, signed char* label) {
+  image_file->seekg(index * rows * cols + 16);
+  image_file->read(pixels, rows * cols);
+  label_file->seekg(index * 4 + 8);
+  label_file->read(label_temp, 4);
+  for (int i = 0; i < 4; i++)
+    *(label+i) = (signed char)*(label_temp+i);
+}
+
+void convert_dataset(const char* image_filename, const char* label_filename,
+        const char* db_filename) {
+  // Open files
+  std::ifstream image_file(image_filename, std::ios::in | std::ios::binary);
+  std::ifstream label_file(label_filename, std::ios::in | std::ios::binary);
+  CHECK(image_file) << "Unable to open file " << image_filename;
+  CHECK(label_file) << "Unable to open file " << label_filename;
+  // Read the magic and the meta data
+  uint32_t magic;
+  uint32_t num_items;
+  uint32_t num_labels;
+  uint32_t rows;
+  uint32_t cols;
+
+  image_file.read(reinterpret_cast<char*>(&magic), 4);
+  magic = swap_endian(magic);
+  CHECK_EQ(magic, 2051) << "Incorrect image file magic.";
+  label_file.read(reinterpret_cast<char*>(&magic), 4);
+  magic = swap_endian(magic);
+  CHECK_EQ(magic, 2050) << "Incorrect label file magic.";
+  image_file.read(reinterpret_cast<char*>(&num_items), 4);
+  num_items = swap_endian(num_items);
+  label_file.read(reinterpret_cast<char*>(&num_labels), 4);
+  num_labels = swap_endian(num_labels);
+  CHECK_EQ(num_items, num_labels);
+  image_file.read(reinterpret_cast<char*>(&rows), 4);
+  rows = swap_endian(rows);
+  image_file.read(reinterpret_cast<char*>(&cols), 4);
+  cols = swap_endian(cols);
+
+  // Open leveldb
+  leveldb::DB* db;
+  leveldb::Options options;
+  options.create_if_missing = true;
+  options.error_if_exists = true;
+  leveldb::Status status = leveldb::DB::Open(
+      options, db_filename, &db);
+  CHECK(status.ok()) << "Failed to open leveldb " << db_filename
+      << ". Is it already existing?";
+
+  char* label_temp = new char[4];  // label for unsigned char*
+  signed char* label_i = new signed char[4];  // label for triplet
+  signed char* label_j = new signed char[4];
+  signed char* label_k = new signed char[4];
+  signed char* label_l = new signed char[4];  // label for pair wise
+  signed char* label_m = new signed char[4];
+  char* pixels = new char[5 * rows * cols];
+  const int kMaxKeyLength = 10;
+  char key[kMaxKeyLength];
+  std::string value;
+
+  caffe::Datum datum;
+  datum.set_channels(5);  // one channel for each image in the triplet and pair
```

Comment:
I had a try on Caffe as you and Manuele described and it could pass the index about class label and pose label for use in the triplet loss layer indeed, so I think it is also able to modify the codes as forwarding a set of samples and their labels of a batch, but I think the problem is that although the feature extarction process is more efficient than ever because some duplex samples will not be forwards many times, but there will be a time consuming in the triplet and pair regrouping process in the triplet and this grouping process could be set at first before we use Caffe for training. In the training process I set each batch used for accumulating loss including all classes and the poses also differs much, so in fact there are not many samples which is duplex with others.  So there I use another idea which Manuele informed that as we use 1 pair wise set and 3 triplet set as a basic unit to calculate a single loss and accumulate each single loss together to get the loss for a turn of forward passing process with a batch. My previous codes will use 11 blobs for this process(1 pair and 3 triplet), but in the new process there will be 5 blobs instead(1 reference sample, 1 with the closest pose, 1 with larger pose and 2 different class), so the pair and 3 triplet will be regrouped as reference+closest pose; reference+closest pose+different class1; .reference+closest pose+different class2; reference+closest pose+larger pose;  And I have also a question that is a loss could be represented as the sum of the loss from the 1 pair and the 3 triplet? or they should be calculated seperately? I think if I use them as a unit it will be both considering class and the pose at the same time. @pwohlhart 

And I find that if I use the 'number' channel to assemble all the samples(5 samples as a unit), the memory will be the same as I use the 'channel' channel to assemble all the samples and still use the slice layer to set them apart in the forward process and collect them again in the loss layer in training stage, I think they will show the same performence on memory, the codes till now is based on the channel seperation method on my codes. What do you think about it Manuele? @mtamburrano